src/scvi/external/decipher/_components.py

```diff
@@ -0,0 +1,102 @@
+from collections.abc import Sequence
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+
+class ConditionalDenseNN(nn.Module):
+    """Dense neural network with multiple outputs, optionally conditioned on a context variable.
+
+    (Derived from pyro.nn.dense_nn.ConditionalDenseNN with some modifications [1])
+
+    Parameters
+    ----------
+    input_dim
+        Dimension of the input
+    hidden_dims
+        Dimensions of the hidden layers (excluding the output layer)
+    output_dims
+        Dimensions of each output layer
+    context_dim
+        Dimension of the context input.
+    deep_context_injection
+        If True, inject the context into every hidden layer.
+        If False, only inject the context into the first hidden layer
+        (concatenated with the input).
+    activation
+        Activation function to use between hidden layers (not applied to the outputs).
+        Default: torch.nn.ReLU()
+    """
+
+    def __init__(
+        self,
+        input_dim: int,
+        hidden_dims: Sequence[int],
+        output_dims: Sequence = (1,),
+        context_dim: int = 0,
+        deep_context_injection: bool = False,
+        activation=torch.nn.ReLU(),
+    ):
+        super().__init__()
+
+        self.input_dim = input_dim
+        self.context_dim = context_dim
+        self.hidden_dims = hidden_dims
+        self.output_dims = output_dims
+        self.deep_context_injection = deep_context_injection
+        self.n_output_layers = len(self.output_dims)
+        self.output_total_dim = sum(self.output_dims)
+
+        # The multiple outputs are computed as a single output layer, and then split
+        indices = np.concatenate(([0], np.cumsum(self.output_dims)))
+        self.output_slices = [slice(s, e) for s, e in zip(indices[:-1], indices[1:], strict=False)]
+
+        # Create masked layers
+        deep_context_dim = self.context_dim if self.deep_context_injection else 0
+        layers = []
+        batch_norms = []
+        if len(hidden_dims):
+            layers.append(torch.nn.Linear(input_dim + context_dim, hidden_dims[0]))
+            batch_norms.append(nn.BatchNorm1d(hidden_dims[0]))
+            for i in range(1, len(hidden_dims)):
+                layers.append(
+                    torch.nn.Linear(hidden_dims[i - 1] + deep_context_dim, hidden_dims[i])
+                )
+                batch_norms.append(nn.BatchNorm1d(hidden_dims[i]))
+
+            layers.append(
+                torch.nn.Linear(hidden_dims[-1] + deep_context_dim, self.output_total_dim)
+            )
+        else:
+            layers.append(torch.nn.Linear(input_dim + context_dim, self.output_total_dim))
+
+        self.layers = torch.nn.ModuleList(layers)
```

Comment:
Is there any advantage here from ModuleList over just the list?

Reply:
Module list ensures the element parameters are properly registered. If they are just in a list and the modules arent direct parameters of the nn.Module class I think it doesnt get picked up by the optimizer